README.md
# gostringsr2

gostringsr2 extracts strings from a Go binary using radare2.

Tested with radare2 3.7.0, Python 3.7, r2pipe 1.4.1, on OS X and Linux.

Tested on Go binaries with architectures: x86 (32 and 64 bit), ARM (32 and 64 bit).

Tested on Go binaries with file formats: ELF (Linux), Mach-O (OS X), PE (Windows).

v1.0.0 code walkthrough - https://carvesystems.com/news/reverse-engineering-go-binaries-using-radare-2-and-python/

## Installation

1. Install radare2.
1. Install gostringsr2 into your Python3 (virtual) environment

From Github:

```
pip install git+https://github.com/carvesystems/gostringsr2
```

or

Locally:

```
git clone https://github.com/carvesystems/gostringsr2
pip install -e gostringsr2
```

## Usage

```
Usage: gostringsr2 [OPTIONS] FILE

Options:
  -n INTEGER  minimum length, default=4
  -v          verbose
  -u          utf8 encoding instead of ascii
  -s TEXT     save output as r2 script; load in r2 with: . [script-file]
  --help      Show this message and exit.
```

## Example

Sample Go file:

```
$ cat <<SOURCE > helloworld.go
package main

func main() {
    print("hello world, how are you today?\n")
}
SOURCE

$ go build helloworld.go

$ ./helloworld
hello world, how are you today?
```

### Basic output:

Find ASCII strings of at least length 8:

```
$ gostringsr2 -n 8 helloworld|grep -B5 -A5 hello
bad write barrier buffer bounds
call from within the Go runtime
casgstatus: bad incoming values
checkmark found unmarked object
entersyscallblock inconsistent 
hello world, how are you today?
inserting span already in treap
internal error - misuse of itab
non in-use span in unswept list
pacer: sweep done at heap size 
resetspinning: not a spinning m
```

### Verbose output:

Shows debug messages and each string's virtual address and (decoded) length.

```
$ gostringsr2 -v -n 8 helloworld|grep -B5 -A5 hello
Loading file into r2: helloworld
file: helloworld
size: 1083 KB
executable: mach0
language: c
architecture: 64-bit x86
os: macos
stripped: False

Locating string table...
String table at 0x106cf20 thru 0x10713a2
Retrieving cross references...
Limited cross-ref check from 0x1001000 to 0x104ead0
Locating string references...
Retrieved 774 references to the string table
Found strings: 631
0x106f9c3 : [31] : bad write barrier buffer bounds
0x106f9e2 : [31] : call from within the Go runtime
0x106fa01 : [31] : casgstatus: bad incoming values
0x106fa20 : [31] : checkmark found unmarked object
0x106fa3f : [31] : entersyscallblock inconsistent 
0x106fa5e : [31] : hello world, how are you today?
0x106fa7d : [31] : inserting span already in treap
0x106fa9c : [31] : internal error - misuse of itab
0x106fabb : [31] : non in-use span in unswept list
0x106fada : [31] : pacer: sweep done at heap size 
0x106faf9 : [31] : resetspinning: not a spinning m
```


### r2 script output

Writes an r2 script that creates:

1. A string reference ("axs") to the string at each code locations
1. A comment ("CCu") at each code reference, `([string length]) "[first 50 characters of the string]"`
1. A flag in the strings flag space starting with `str.go.[first 20 chars of the string]`

```
$ gostringsr2 -s helloworld.r2 -v -n 8 -helloworld|grep hello
Loading file into r2: helloworld
file: helloworld
size: 1083 KB
executable: mach0
language: c
architecture: 64-bit x86
os: macos
stripped: False

Locating string table...
String table at 0x106cf40 thru 0x1071403
Retrieving cross references...
Limited cross-ref check from 0x1001000 to 0x104eaf0
Locating string references...
Retrieved 775 references to the string table
Found strings: 632
+ r2 script written to hello.r2. Load in r2 with '. [scriptfile]'
0x106fbf7 : [32] : hello world, how are you today?


$ r2 helloworld
 -- It's not you, it's me.
[0x0104a4d0]> . hello.r2
[0x0104a4d0]> axt 0x106fbf7
(nofunc); (32) "hello world, how are you today?//" 0x104ea42 [STRING] lea rax, str.go.hello_world__how_are
[0x0104a4d0]> pd 6 @0x104ea42
            0x0104ea42      488d05ae1102.  lea rax, str.go.hello_world__how_are ; (32) "hello world, how are you today?//"
            0x0104ea49      48890424       mov qword [rsp], rax
            0x0104ea4d      48c744240820.  mov qword [rsp + 8], 0x20
            0x0104ea56      e87557fdff     call sym.runtime.printstring
            0x0104ea5b      e8c04efdff     call sym.runtime.printunlock
            0x0104ea60      e83b4efdff     call sym.runtime.printlock
[0x0104a4d0]>
```


CHANGELOG.md
# Changelog

## 1.1.2

- Validate file is elf/pe/mach0 or throw an error
- Rearrange logic so that non-ARM/x86 architectures may work
- Add some docs to GoStringsR2
- Open r2pipe with `-2` flag to kill stderr

## 1.1.1

- Add `-s` option to generate an r2 script that can be loaded into radare2 afterwards

## 1.1.0

- Use `p8` instead of `pr` to get raw data (#1)
- Fix bug in is_a_string_ref() (#1)
- Use quietr2pipe to kill r2's stderr during run()
- Some refactoring, debug output, optimizing cross-ref boundaries for r2

## 1.0.0

- Initial release


gostringsr2/cli.py
import sys
import click
import re
from os import path
from . import GoStringsR2, GoStringsR2Error


def printe(*args, **kwargs):
    print(*args, file=sys.stderr, flush=True, **kwargs)


@click.command()
@click.argument("file")
@click.option(
    "-n", "length", is_flag=False, default=4, help="minimum length, default=4"
)
@click.option("-v", "verbose", is_flag=True, help="verbose")
@click.option("-u", "utf8", is_flag=True, help="utf8 encoding instead of ascii")
@click.option(
    "-s", "r2script", help="save output as r2 script; load in r2 with: . [script-file]"
)
def main(file, length, verbose, utf8, r2script):

    if not path.isfile(file):
        printe("invalid file {}".format(file))
        return 1

    g = GoStringsR2(file, verbose)

    try:
        g.load()

        encoding = "ascii"
        if utf8:
            encoding = "utf8"

        go_strings = g.get_strings(length, encoding)
        # array of [address, decoded length, string, byte length, list of code refs]

        if r2script is not None:
            r2scriptfile = open(r2script, "w")
            r2scriptfile.writelines(
                [
                    "fs strings\n",
                    "e asm.comments = false\n",  # The big Go string gets in the way
                    "e asm.usercomments = true\n",
                ]
            )
        else:
            r2scriptfile = None

        for go_string in go_strings:
            s_addr, s_len, s_val, s_binlen, s_refs = go_string

            # get rid of "binary" chars before printing, otherwise pipes to grep are unhappy
            s_val = re.sub("[\x00\x08]", "", s_val)

            if verbose:
                print("0x{:x} : [{}] : {}".format(s_addr, s_len, s_val))
            else:
                print(s_val)

            if r2scriptfile is not None:
                r2scriptfile.writelines(g.get_r2_script_for_string(go_string, encoding))

        if r2scriptfile is not None:
            printe(
                "+ r2 script written to {}. Load in r2 with '. [scriptfile]'".format(
                    r2script
                )
            )
            r2scriptfile.close()

        g.kill()

        return 0

    except GoStringsR2Error as err:
        printe("gostringsr2 error: {}".format(err))


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


gostringsr2/r2piper.py
import r2pipe
import os
import time
import io
import sys
from subprocess import Popen, PIPE


class open(r2pipe.open_sync.open):
    def __init__(self, filename="", flags=[], radare2home=None):
        super(open, self).__init__("", flags)
        if filename:
            self._cmd = self._cmd_process
            if radare2home is not None:
                if not os.path.isdir(radare2home):
                    raise Exception(
                        "`radare2home` passed is invalid, leave it None or put a valid path to r2 folder"
                    )
                r2e = os.path.join(radare2home, "radare2")
            else:
                r2e = "radare2"
                if os.name == "nt":
                    # avoid errors on Windows when subprocess messes with name
                    r2e += ".exe"
            cmd = [r2e, "-q0", filename]
            cmd = cmd[:1] + flags + cmd[1:]
            dnull = io.open(os.devnull, "w")
            try:
                self.process = Popen(
                    cmd, shell=False, stdin=PIPE, stdout=PIPE, stderr=dnull, bufsize=0
                )
            except:
                raise Exception("ERROR: Cannot find radare2 in PATH")
            self.process.stdout.read(1)  # Reads initial \x00
            # make it non-blocking to speedup reading
            self.nonblocking = True
            if self.nonblocking:
                fd = self.process.stdout.fileno()
                if not self.__make_non_blocking(fd):
                    Exception("ERROR: Cannot make stdout pipe non-blocking")

    """ Override _cmd_process to return bytes instead of a decoded utf8 string """

    def _cmd_process(self, cmd):
        cmd = cmd.strip().replace("\n", ";")
        self.process.stdin.write((cmd + "\n").encode("utf8"))
        r = self.process.stdout
        out = b""
        sys.stdout.flush()  # flush any output before running command, in case thecommand takes a while
        c = 0
        while True:
            if self.nonblocking:
                try:
                    foo = r.read(4096)
                except:
                    continue
            else:
                foo = r.read(1)
            if foo:
                if foo.endswith(b"\0"):
                    out += foo[:-1]
                    break

                out += foo
            else:
                # if there is no any output from pipe this loop will eat CPU, probably we have to do micro-sleep here
                if self.nonblocking:
                    time.sleep(0.001)
                    c += 1
                    if c % 50 == 0:
                        print(
                            ["_", "-", "+", "-"][c // 50 % 4],
                            end="",
                            file=sys.stderr,
                            flush=True,
                        )

        if c >= 50:
            print("", file=sys.stderr, flush=True)
        return out


gostringsr2/quietr2pipe.py
import r2pipe
import os
import time
import io
import sys
from subprocess import Popen, PIPE


class open(r2pipe.open_sync.open):
    def __init__(self, filename="", flags=[], radare2home=None):
        if filename:
            self._cmd = self._cmd_process
            if radare2home is not None:
                if not os.path.isdir(radare2home):
                    raise Exception(
                        "`radare2home` passed is invalid, leave it None or put a valid path to r2 folder"
                    )
                r2e = os.path.join(radare2home, "radare2")
            else:
                r2e = "radare2"
                if os.name == "nt":
                    # avoid errors on Windows when subprocess messes with name
                    r2e += ".exe"
            cmd = [r2e, "-q0", filename]
            cmd = cmd[:1] + flags + cmd[1:]
            dnull = io.open(os.devnull, "w")
            try:
                self.process = Popen(
                    cmd, shell=False, stdin=PIPE, stdout=PIPE, stderr=dnull, bufsize=0
                )
            except:
                raise Exception("ERROR: Cannot find radare2 in PATH")
            self.process.stdout.read(1)  # Reads initial \x00
            # make it non-blocking to speedup reading
            self.nonblocking = True
            if self.nonblocking:
                fd = self.process.stdout.fileno()
                if not self.__make_non_blocking(fd):
                    Exception("ERROR: Cannot make stdout pipe non-blocking")


gostringsr2/gostringsr2.py
#!/usr/bin/env python3

import sys
import json
import binascii
import re
import base64

import r2pipe


class GoStringsR2Error(RuntimeError):
    pass


class GoStringsR2:

    SUPPORTED_ARCHS = ["arm", "x86"]
    SUPPORTED_BINTYPES = ["elf", "pe", "mach0"]

    def __init__(self, _file, _logging=False):
        """
        Initialize GoStringsR2 with a path to a Binary 

        _file is the path to a file.
        If _logging is True, status messages will be output to standard error.
        """

        self.file = _file
        self.logging = _logging
        self.loaded = False
        self.r2 = None

    def kill(self):
        """
        Closes the r2pipe session
        """

        if self.loaded:
            self.r2.quit()
            self.r2 = None
            self.loaded = False

    def runjson(self, cmd):
        """
        Executes an r2 command that returns a JSON dictionary.
        """

        return self.r2.cmdj(cmd)

    def run(self, cmd):
        """
        Executes an r2 command
        """

        return self.r2.cmd(cmd)

    def load(self):
        """
        Opens the r2pipe session.

        GoStringsR2Error may be thrown if there is an error loading.
        """

        self.log("Loading file into r2: {}".format(self.file))
        self.r2 = r2pipe.open(self.file)
        self.data = {}
        self.data["info"] = self.runjson("ij")
        if "bin" not in self.data["info"]:
            raise GoStringsR2Error("r2 could not parse the binary")

        self.arch = self.data["info"]["bin"]["arch"]
        self.bintype = self.data["info"]["bin"]["bintype"]
        self.bits = self.data["info"]["bin"]["bits"]
        self.binos = self.data["info"]["bin"]["os"]

        if self.bintype not in ["elf", "mach0", "pe"]:
            raise GoStringsR2Error(
                "bintype {} not supported by gostringsr2. Supported: {}".format(
                    self.bintype, GoStringsR2.SUPPORTED_BINTYPES
                )
            )
        if self.arch not in ["arm", "x86"]:
            self.log("warning: arch {} may not fully work".format(self.arch))

        self.data["symbols"] = self.runjson("isj")
        self.data["sections"] = self.runjson("iSj")

        self.loaded = True

        self.log(self.file_info())

    def file_info(self):
        """
        Returns a descriptive string of the loaded binary.
        """

        if self.loaded:
            return (
                "file: {}\n"
                "size: {} KB\n"
                "executable: {}\n"
                "language: {}\n"
                "architecture: {}-bit {}\n"
                "os: {}\n"
                "stripped: {}\n".format(
                    self.data["info"]["core"]["file"],
                    self.data["info"]["core"]["size"] // 1024,
                    self.data["info"]["bin"]["bintype"],
                    self.data["info"]["bin"]["lang"],
                    self.data["info"]["bin"]["bits"],
                    self.data["info"]["bin"]["arch"],
                    self.data["info"]["bin"]["os"],
                    self.data["info"]["bin"]["stripped"],
                )
            )

        return "file: <none>"

    def get_string_table_symbols(self, rdata):
        """
        Returns a dictionary with the raw data from the string table, as found by referencing Go symbols in the provided rdata dictionary.
        """

        g_str = self.find_symbol("go.string.*")
        g_func = self.find_symbol("go.func.*")
        if g_str is not None and g_func is not None:
            g_str["tabsize"] = g_func["vaddr"] - g_str["vaddr"]
            startaddr = g_str["vaddr"] - rdata["vaddr"]
            endaddr = startaddr + g_str["tabsize"]
            g_str["table"] = rdata["data"][startaddr:endaddr]
            return g_str

        return None

    def get_rodata_section(self):
        """
        Returns the first read-only data section in the binary.
        """

        if self.bintype == "elf":
            sname = ".rodata"
        elif self.bintype == "mach0":
            sname = ".__TEXT.__rodata"
        elif self.bintype == "pe":
            sname = ".rdata"
        return self.get_section_data(sname)

    def get_code_section(self):
        """
        Returns the first text/code section in the binary.
        """

        if self.bintype in ["elf", "pe"]:
            return self.get_section_info(".text")
        elif self.bintype == "mach0":
            return self.get_section_info(".__TEXT.__text")
        return None

    def get_string_table_search(self, rdata):
        """
        Returns a dictionary with the raw data from the string table, as found via searching in the provided rdata dictionary.
        """

        self.log("Searching for string table")
        if rdata is not None:
            str_start, str_size = self._find_longest_string(rdata["data"])

            if str_size > 0:
                g_str = {"vaddr": rdata["vaddr"] + str_start, "tabsize": str_size}
                startaddr = g_str["vaddr"] - rdata["vaddr"]
                endaddr = startaddr + g_str["tabsize"]
                g_str["table"] = rdata["data"][startaddr:endaddr]

                return g_str

        return None

    def _find_longest_string(self, bindata):
        off = 0
        this_off = 0
        longest_off = 0
        longest_size = 0

        binlength = len(bindata)
        while off < binlength:
            b = bindata[off : off + 2]
            # Basically, terminate a "string" if 2 null bytes are seen. Seems to work for the most part.
            if b == b"\x00\x00":
                this_size = off - this_off
                if this_size > 0:
                    if this_size > longest_size:
                        longest_off = this_off
                        longest_size = this_size
                this_off = off + 2
            else:
                this_size = off - this_off
                if this_size > 0:
                    if this_size > longest_size:
                        longest_off = this_off
                        longest_size = this_size
            off += 2

        if (off - this_off) > longest_size:
            longest_off = this_off
            longest_size = off - this_off

        if longest_size > 0:
            return (longest_off, longest_size)

        return (None, 0)

    def get_string_table(self):
        """
        Returns the string table either via symbols or via searching.
        """

        rodata = self.get_rodata_section()
        stab_sym = self.get_string_table_symbols(rodata)
        stab_sym = (
            stab_sym if stab_sym is not None else self.get_string_table_search(rodata)
        )

        if stab_sym is None:
            return None
        else:
            strtab_start = stab_sym["vaddr"]
            strtab_end = strtab_start + stab_sym["tabsize"]
            self.log(
                "String table at 0x{:x} thru 0x{:x}".format(strtab_start, strtab_end)
            )
            strtab = {
                "startaddr": strtab_start,
                "endaddr": strtab_end,
                "data": stab_sym["table"],
            }
            return strtab

    def find_symbol(self, symbol_name):
        """
        Returns a symbol in the binary as a dictionary, as retrieved with r2.
        """

        for sym in self.data["symbols"]:
            if sym.get("name", "") == symbol_name:
                return sym
        return None

    def get_cross_refs(self):
        """
        Performs the cross-references search and returns results in r2 quiet/human-readable format.
        """

        xrefs = None

        # Only check .text; other executable sections may get searched otherwise
        # If more than one .text section exists, changeme
        code_section = self.get_code_section()
        if code_section is not None:
            c_start = code_section["vaddr"]
            c_end = c_start + code_section["size"]
            self.log(
                "Limited cross-ref check from 0x{:x} to 0x{:x}".format(c_start, c_end)
            )
            self.run("e search.from=0x{:x}".format(c_start))
            self.run("e search.to=0x{:x}".format(c_end))

        cross_ref_cmd = "/ra"
        # Use ESIL analysis for non-x86 architectures
        if self.arch != "x86":
            cross_ref_cmd = "aae"
        # send stderr from r2 to /dev/null to hide r2's address progress
        self.run("{} 2>/dev/null".format(cross_ref_cmd))
        xrefs = self.run("axq")
        return xrefs

    def get_section_info(self, section_name):
        """
        Returns the section info of section_name, as retrieved with r2.
        """

        for secobj in self.data["sections"]:
            if secobj["name"].endswith(section_name):
                return secobj
        return None

    def get_section_data(self, section_name):
        """
        Returns a dictionary containing the raw binary data of the requested section.
        """

        secobj = self.get_section_info(section_name)
        if secobj is not None:
            s_base = secobj["vaddr"]
            s_size = secobj["vsize"]
            rdsize = 4096
            i = 0
            sdata = b""
            while s_size > 0:
                c = "p8 {} @0x{:x}".format(min(rdsize, s_size), s_base + i * rdsize)
                sdat = self.run(c).strip()
                sdata += binascii.unhexlify(sdat)
                i += 1
                s_size -= rdsize

            return {"name": section_name, "vaddr": s_base, "data": sdata}
        return None

    def find_strings(self, minlength, encoding, refs, tablebase, tabledata):
        """
        Processes cross-references and returns a list of string objects.

        Each string object in the returned list is a list specifying address, decoded length, decoded string value, string size in bytes, list of code references to this string.
        """

        # refs.keys() = dest address, refs.values() = list of source addresses
        refs_addrs = sorted(refs.keys(), reverse=True)

        all_strings = []
        for r in refs_addrs:
            # r = virtual addr of a string
            # subtract vaddr of section to get offset into
            r_offset = r - tablebase
            if len(all_strings) > 0:
                last_ref = all_strings[len(all_strings) - 1][0] - tablebase
                r_end_offset = last_ref
            else:
                r_end_offset = len(tabledata)

            r_str = tabledata[r_offset:r_end_offset].decode(encoding, errors="ignore")
            decoded_len = len(r_str)
            all_strings.append(
                [
                    tablebase + r_offset,
                    decoded_len,
                    r_str,
                    r_end_offset - r_offset,
                    refs[r],
                ]
            )

        # filter all_strings by length requirement, then reverse order
        # since all_strings started at the end
        return list(reversed([s for s in all_strings if s[1] >= minlength]))

    def _is_a_string_ref(
        self, src_addr, dst_addr, strtab_addr, strtab_endaddr, code_section
    ):
        if dst_addr >= strtab_addr and dst_addr < strtab_endaddr:
            if code_section is None:
                return True
            else:
                return src_addr >= code_section["vaddr"] and src_addr < (
                    code_section["vaddr"] + code_section["size"]
                )

        return False

    def process_xrefs(self, xrefs, strtab_start, strtab_end):
        """
        Filters cross-references to only references to the string table.

        xrefs is data returned by r2 from "axq" (quiet/human-readable format), strtab_start/end specify the addresses of the string table.
        """

        str_refs = {}

        code_section = self.get_code_section()

        # 0x01640839 -> 0x016408a9  CALL
        for line in xrefs.split("\n"):
            lparts = line.split(" ")
            # 0 = src, 1= arrow, 2 = dst, 3=empty, 4=type
            if len(lparts) == 5:
                r_src = int(lparts[0], 16)
                r_dst = int(lparts[2], 16)
                if self._is_a_string_ref(
                    r_src, r_dst, strtab_start, strtab_end, code_section
                ):
                    if r_dst in str_refs.keys():
                        str_refs[r_dst].append(r_src)
                    else:
                        str_refs[r_dst] = [r_src]

        return str_refs

    def log(self, log_msg, *args, **kwargs):
        if self.logging:
            print("\033[92m" + log_msg + "\033[0m", *args, **kwargs, file=sys.stderr)

    def get_strings(self, minlength, encoding="ascii"):
        """
        Perform the string search, returning string objects.

        minlength specifies the minimum length of a string that will be returned.
        encoding specifies how bytes should be decoded, with "ascii" being the default.
        """

        ret_strings = []

        self.log("Locating string table...")
        strtab = self.get_string_table()
        if strtab is None:
            raise GoStringsR2Error("couldn't find the Go string table in the binary")

        self.log("Retrieving cross references...")
        xrefs = self.get_cross_refs()
        if xrefs is None:
            raise GoStringsR2Error("r2 returned no cross-references")

        self.log("Locating string references...")
        str_refs = self.process_xrefs(xrefs, strtab["startaddr"], strtab["endaddr"])

        self.log("Retrieved {} references to the string table".format(len(str_refs)))
        if len(str_refs):
            ret_strings = self.find_strings(
                minlength, encoding, str_refs, strtab["startaddr"], strtab["data"]
            )

        self.log("Found strings: {}".format(len(ret_strings)))
        return ret_strings

    def get_r2_script_for_string(self, go_string, encoding):
        """
        Returns a list of r2 commands that create comments at code locations where strings are referenced.
        """

        s_addr, s_decoded_len, s_value, s_bin_len, s_refs = go_string
        script = []

        # sanitized flag name, up to 20 alpha numeric chars
        san_str = re.sub("[^a-zA-Z0-9]", "_", s_value)[:20]

        # comment will be up to 50 chars; replace newlines with double slash //
        trunc_str = '({}) "{}{}"'.format(
            s_decoded_len,
            re.sub("[\n]", "//", s_value[:50]),
            "..." if s_decoded_len > 50 else "",
        )
        # ...but base64 encoded because r2 doesn't like ; chars
        trunc_str = base64.b64encode(trunc_str.encode(encoding)).decode("ascii")

        # create flag for the string value, use bin length for r2
        script.append("f str.go.{} {} @0x{:x}\n".format(san_str, s_bin_len, s_addr))
        # create string reference and comment at each reference containing a snippet of the string
        for r in s_refs:
            script.append("axs 0x{:x} @0x{:x}\n".format(s_addr, r))
            script.append("CCu base64:{} @0x{:x}\n".format(trunc_str, r))

        return script
